### PR TITLE
Validation test

### PR DESCRIPTION
**Pull request by malikhaiderkhan**
In this project I implemented:

- Add validations for user and post models.
- Add unit specs for all models methods and validations.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    subject.posts_counter = 5
+    expect(subject).to be_valid
+
+    subject.posts_counter = 'abc'
+    expect(subject).to_not be_valid
+  end
+
+  it 'expect recent posts to be 0 ' do
+    expect(subject.three_recent_posts.length).to eq(0)
+  end
```
Comment (Ghiftee) [inline spec/models/user_spec.rb]: - [ ] Excellent work you have done creating these tests 👍🏽 . Just a few changes. Kindly wrap the `it` block inside a `describe 'validations'` block. Aside from the fact that it is best practice, using `describe` blocks helps make the structure of your test suite clear. It provides context for the tests and makes it explicit what part of the model you are testing. Your code should look something like this:
```ruby
require 'rails_helper'

RSpec.describe User, type: :model do
  subject(:user) { User.new(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.') }

  before { user.save }

  describe 'validations' do
    it 'name should be present' do
      another_user = User.create(photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
      expect(another_user).to_not be_valid
    end

    it 'post_count should be a non-negative integer' do
      user.posts_counter = -1
      expect(user).to_not be_valid

      user.posts_counter = 0
      expect(user).to be_valid

      user.posts_counter = 5
      expect(user).to be_valid

      user.posts_counter = 'abc'
      expect(user).to_not be_valid
    end
  end

  describe 'recent posts' do
    it 'expect recent posts to be 0' do
      expect(user.three_recent_posts.length).to eq(0)
    end
  end
end

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  it 'likes_counter should be non negative integer' do
+    subject.likes_counter = -1
+    expect(subject).to_not be_valid
+
+    subject.likes_counter = 0
+    expect(subject).to be_valid
+
+    subject.likes_counter = 5
+    expect(subject).to be_valid
+  end
```
Comment (Ghiftee) [inline spec/models/post_spec.rb]: - [ ] Excellent work you have done creating these tests 👍🏽 . Just a few changes. Kindly wrap the `it` block inside a `describe 'validations'` block. Aside from the fact that it is best practice, using `describe` blocks helps make the structure of your test suite clear. It provides context for the tests and makes it explicit what part of the model you are testing. Your code should look something like this:
```ruby
require 'rails_helper'

RSpec.describe Post, type: :model do
  let(:first_user) { User.create(name: 'John') }

  subject(:post) { Post.new(author: first_user, title: 'Hello', text: 'This is my first post') }

  before { post.save }

  describe 'validations' do
    it 'title should be present' do
      post.title = nil
      expect(post).to_not be_valid
    end

    it 'title should be less than or equal to 250' do
      post.title = 'a' * 300
      expect(post).to_not be_valid
    end

    it 'comments_counter should be non-negative integer' do
      post.comments_counter = -1
      expect(post).to_not be_valid

      post.comments_counter = 0
      expect(post).to be_valid

      post.comments_counter = 5
      expect(post).to be_valid
    end

    it 'likes_counter should be non-negative integer' do
      post.likes_counter = -1
      expect(post).to_not be_valid

      post.likes_counter = 0
      expect(post).to be_valid

      post.likes_counter = 5
      expect(post).to be_valid
    end
  end

  describe 'associations' do
    it 'should update posts counter' do
      Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
      Post.create(author: first_user, title: 'Hello', text: 'This is my 2nd post')
      expect(first_user.posts_counter).to eq(3)
    end

    it 'should return five recent comments' do
      post.save
      post.comments.create(user: first_user, text: 'Comment 1')
      comment2 = post.comments.create(user: first_user, text: 'Comment 2')
      comment3 = post.comments.create(user: first_user, text: 'Comment 3')
      comment4 = post.comments.create(user: first_user, text: 'Comment 4')
      comment5 = post.comments.create(user: first_user, text: 'Comment 5')
      comment6 = post.comments.create(user: first_user, text: 'Comment 6')

      recent_comments = post.five_recent_comments

      expect(recent_comments).to eq([comment6, comment5, comment4, comment3, comment2])
    end
  end
end

**Review by Ghiftee — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️♻️
Hello @malikhaiderkhan 👋🏽 

Good job working on the project so far! 
To highlight 👍🏽 
✔️ Tests pass
✔️ Required tests are added
✔️ README is professional

However, there are some issues that you need to work on to go to the next project but you are almost there!



## Required Changes ♻️
_Check the comments under the review 👇🏽 _

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Review by Felix45 — APPROVED** [review]
# Approved :green_circle: 
Hi @malikhaiderkhan, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights :heavy_check_mark: 
- [x] All models have been set correctly :white_check_mark: 
- [x] All migrations run as expected :white_check_mark: 
- [x] All model validations are set up correctly :white_check_mark: 
- [x] All model specs pass with green checkmarks :white_check_mark:   
- [x] Pull request has a descriptive title and summary :white_check_mark: 
- [x] Repository has a professional readme file :white_check_mark:  

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**